### ISSUE TITLE
A web clipper

### ISSUE THREAD

**Issue by ylian** [HappenApps/Quiver#100]
A web clipper can be useful for clipping blog posts, tutorials, etc.

Evernote's web clipper is one of the best features of Evernote. It has a few options: full article, simplified article, URL. I personally use simplified article most often.

I think Quiver can do even better. It should also provide an option for Markdown, using [unmarked](https://github.com/ylian/unmarked). So it's a two-step process, use readability.js to clean up a web page, then use unmarked to convert to Markdown (similar to http://heckyesmarkdown.com/,  but all client-side). Quiver can even convert fenced code blocks into code cells, with proper language detection. This will be really neat for clipping programming tutorials or blog posts.

This applies to both OS X and iOS.

**Comment by LeonardoGentile** [comment]
:+1: 
Evernote clipper will automatically upload the content to the user online account. How Quiver will handle the clipped content, where will it be posted? Are you planning already an online service?

**Comment by ylian** [comment]
No, no online service. The web clipper clips the content, saves it to the system clipboard, then opens a custom URL like `quiver://notes/create` to invoke Quiver. Quiver then processes the URL and clipboard data and creates a new note. This is the standard practice for many apps, such as nvAlt and DEVONote.

I might add a menubar app to facilitate this process. So you don't have to open the Quiver main app every time. The menubar app would be able to create a Quiver note from clipboard data.

**Comment by LeonardoGentile** [comment]
Cool :+1:

**Comment by ylian** [comment]
Spent a few days working on my own version of Readability: https://github.com/ylian/declutter. Still work in progress, but already significantly faster than other alternatives (2x to 10x faster), and can handle most pages.

The goal is to be able to run `declutter` then `unmarked` to convert a complex HTML page (e.g., a tutorial site) into simple, clean Markdown. I want this to be fast on iOS as well, so speed is important.

**Comment by preflight** [comment]
This would be a really valuable feature; thanks for considering it

**Comment by ghost** [comment]
OH SWEET. Really looking forward to this.

**Comment by aamnah** [comment]
+1 for a web clipper. YES!

**Comment by JMichaelTX** [comment]
Having a web clipper for Quiver would be huge.  I do a ton of web clipping.
Any idea on when this might be available?

**Comment by ylian** [comment]
@JMichaelTX High on the list.

**Comment by JMichaelTX** [comment]
:+1:

**Comment by bborysenko** [comment]
:+1:

**Comment by ChristinWhite** [comment]
@ylian I noticed you mentioned web clippings on your [What is Quiver?](http://yaoganglian.com/2015/12/06/What-is-Quiver/) post, has the clipper moved ahead and I missed it or did you mean manually doing it for now?

Thanks!

**Comment by ylian** [comment]
@christopherdwhite No you didn't miss it. Still in development.

**Comment by JMichaelTX** [comment]
While we're waiting for the cool clipper that I'm sure @ylian will provide, I have written my own simple clipper using Keyboard Maestro.  If there's enough interest, I'll clean it up for public consumption and post.  It works pretty good, but there are a few occasional glitches.

**Comment by shartf** [comment]
@JMichaelTX

Would you care to share the workflow for KM? Thanks!

**Comment by JMichaelTX** [comment]
> Would you care to share the workflow for KM? Thanks!

Sure.  It might take me a few days to cleanup my macro for public consumption.  If I have not posted back here within a week, feel free to post/ping me as a reminder.  :wink:

**Comment by shartf** [comment]
@JMichaelTX ping :)

**Comment by JMichaelTX** [comment]
> Would you care to share the workflow for KM? Thanks!

Sorry it took so long.  Here it is:
[Macro: [QUIVER] Copy Selection to New Quiver Note](https://forum.keyboardmaestro.com/t/macro-quiver-copy-selection-to-new-quiver-note/2757)

Please read the header comments carefully, as there are several KM Actions you will want to change to suit your needs.  

This KM macro seems to be working fine, at least for me, but then I wrote it.  :smile: 
So you may find bugs or issues, or have suggestions for improvement.  If so, please post in the KM Macro thread above.

**Comment by moubi** [comment]
@ylian any progress on the web clipper? Is it still high in the list?

**Comment by ylian** [comment]
@moubi About 80% complete. Many things will be released in the next two months.

**Comment by inspiredearth** [comment]
Any updates on the Webclipper?
Thanks...

**Comment by geckofu** [comment]
Looking forward to the webclipper

**Comment by inspiredearth** [comment]
I wouldn't hold your breath for it. From what I can tell, over the past 
year or more, development on this app is very slow, if happening at all. 
Hard to say.
 A web clipper would be great though.

**Comment by aamnah** [comment]
Yeah.. the web clipper has been _in the works_ since April 2015..

**Comment by rmrnc** [comment]
Is there any progress/update on this or at least something resembling a timeline?

**Comment by matchavez** [comment]
If there’s one thing about Quiver, it is a hobby. It is on the edge of being great, but it seems we have what we have.

**Comment by kimaldis** [comment]
FWIW: I've found the most reliable way to clip from web to Quiver is to use the Evernote web clipper, export the clipped note to .enex then import this into Quiver. It nicely brings everything, including images. When I get a moment I'll have a crack at making it more automatic.

**Comment by imcoddy** [comment]
Is there any progress/update on this or at least something resembling a timeline?

Really would love to have it

**Comment by ylian** [comment]
@imcoddy I will have a working version out this year.

**Comment by imcoddy** [comment]
Any updates for this feature? Maybe we are aiming at Chinese New Year perhaps? :)

**Comment by ylian** [comment]
@imcoddy Almost there. There are some cool features that you will like. :)

**Comment by imcoddy** [comment]
Any ETA when we can see it?

**Comment by ylian** [comment]
@imcoddy I am finishing up this one. Give me a couple more days.

**Comment by ylian** [comment]
Quiver Web Clipper 1.0 is released: https://chrome.google.com/webstore/detail/quiver-web-clipper/hcnffmpopoelpggikahccdfenoobjigj

Please make sure to update Quiver to 3.2.4.

**Comment by steverowling** [comment]
This looks great. Any chance you could create a Safari extension version please?

**Comment by AndrewMac** [comment]
Yes, please: a Safari extension would be great. And if you had an option to limit the text to a selection on the page, along with other cool stuff, then I think I might even be prepared to pay for it ;-)

**Comment by ylian** [comment]
@steverowling Sure. Modern browser extensions are easy to port across browsers.

**Comment by waz0wski** [comment]
thanks for releasing this! if possible, please publish an extension for firefox as well.

**Comment by ylian** [comment]
@waz0wski Firefox literally can use the same extension without change, so sure will upload to Firefox Web Store.

**Comment by ylian** [comment]
@AndrewMac Can you elaborate on "clipping selection"? I can think of several user flows, but not sure which one is most intuitive to users.

Flow 1: Select some text from the original page, turn on "Quiver Web Clipper", the selected text will be cleaned up and shown in the clipper's main window, then click on the "Clip to Quiver" button to create a Quiver note with the selected text.

Flow 2: Select some text from the original page, right click and select "Clip to Quiver" from the contextual menu, and a Quiver note with the selected text is created.

Flow 3: Turn on "Quiver Web Clipper", select some text, click on "Clip to Quiver" button, and a Quiver note with the selected text is created.

Flow 4: Something like Evernote Web Clipper's selection feature. Basically you can mouse over sections of original page and choose which section to clip.

**Comment by AndrewMac** [comment]
@ylian Thanks for taking an interest! But I must admit that I've not given the workflow a lot of thought. Basically, I almost never clip a whole page but rather just a single section of text (like a code example, for instance), or several non-contiguous sections. I usually then pop them into quiver — often in different cells, mixing text with code, etc — with a note that links back to the original page.

I think that what I personally would find most intuitive would be to first turn on the Quiver Web Clipper, select some text, and then look for some kind of interface widget to do the clip. So I guess that would be Flow 3. But I've not actually tried the Evernote approach, so that might be okay too. I think I'd trust your judgement on it, though, and to be honest any of those four approaches would be great. I'd just love to be able to clip just a selection.

If it were possible to include non-contiguous selections, though, that would be absolutely fantastic. And if each individual one of those could end up in a different cell that (for me!) would be even better.

And finally, if the clipper were to add a link back to the original page in the first cell I would feel even more wonderful about using Quiver.

But hey, I'll be using it anyway, no matter what ;-)

**Comment by inspiredearth** [comment]
> Quiver Web Clipper 1.0 is released: https://chrome.google.com/webstore/detail/quiver-web-clipper/hcnffmpopoelpggikahccdfenoobjigj
> 
> Please make sure to update Quiver to 3.2.4.

Thank you very much for your work on this. I am very glad to see Quiver is still being actively developed, with such useful features as this being added. I use Quiver a huge amount. I've got over coding and server management related 500 notes in there.

I'll try out the clipper (in Chrome) and let you know my thoughts on possible workflows. I will also await a Firefox version. For your info: A Safari version is of no interest to me, as I use Firefox and Chrome exclusively, due to their excellent handling of web development.

**Comment by inspiredearth** [comment]
I was just thinking... Would there be any advantages to using an OS X service to handle the clipping, etc., rather than various browser plug-ins? I have a number of other apps that handle things that way... such as a recipe clipper, which I access via OS X Services (and is therefore independent of the browser, or whichever other source I might be clipping the text from (e.g., I might be clipping it from a PDF, or some other article, or content in any app on OS X).

**Comment by inspiredearth** [comment]
Having now tested the clipper... here are some thoughts on work flow.

> Flow 1: Select some text from the original page, turn on "Quiver Web Clipper", the selected text will be cleaned up and shown in the clipper's main window, then click on the "Clip to Quiver" button to create a Quiver note with the selected text.
> 
> Flow 2: Select some text from the original page, right click and select "Clip to Quiver" from the contextual menu, and a Quiver note with the selected text is created.
> 
> Flow 3: Turn on "Quiver Web Clipper", select some text, click on "Clip to Quiver" button, and a Quiver note with the selected text is created.
> 
> Flow 4: Something like Evernote Web Clipper's selection feature. Basically you can mouse over sections of original page and choose which section to clip.

I think initially Flow 3 is fine, although currently the clipper pays no attention to the fact I have selected some text. The entire page ends up in the new clip. I was trying it out on here: https://apple.stackexchange.com/questions/175676/how-can-i-automate-pasting-of-password-into-securityagent/348703#348703

Flow 2 would also be fine, but I suspect it will be easier to allow for feature number 1 (below) if Flow 3 is used. Also, in some cases selecting from cleaned text (rather than the original, which might have an advert in the middle of it), which is achieved with Flow 3.

Also, it pays no attention to the code blocks. Meaning, they don't end up in Code format in the resulting clip.

Eventually, however, I would find it useful to have the following features:
1) The clipper will have two buttons, as opposed to he one it has now. One button will create a new note. The second button will append to an existing note (which ever is currently open).

2) The clipper will copy the source URL, and add this to the top of the new clip, as "Source: {URL}", then one line of space, then the beginning of the clip.

3) It would be great if it where possible to do something similar to Flow 4. Where you can select multiple sections in a page of text, and have them clipped into one clip (either a new clip, or appended to existing one, as per point 1 above).

**Comment by inspiredearth** [comment]
Another issue. I tested clicking the Clipper button whilst viewing this Git thread. It displays only the following in the cleaned text:

![image](https://user-images.githubusercontent.com/667395/52106668-844b0f80-2658-11e9-85ca-3c43056f1cb0.png)

**Comment by ylian** [comment]
@AndrewMac @jonathannz Thanks for the feedback. I have a good sense now how to implement "clip selection" based on the feedback.

@jonathannz  Web clipper is different from system services because it can provide extra things such as simplifying the page and displaying it in a Reader View. Appending to a note probably won't be supported, as I aim to handle the most important use cases and don't want to make the clipper too complicated. URL will certainly be added.

@jonathannz Also, the clipper is designed for article pages (a page containing a single article). A page consisting of multiple comments like the current page won't work. But this issue can be alleviated with "clip selection" support.

**Comment by sapjax** [comment]
Hi @ylian 
Right now,  I prefer to copy the HTML content manually, cause the clipper lost the format of origin HTML.

**Comment by ylian** [comment]
@sapjax That's fine. The web clipper's main use case is to clip simplified article content. If you want to keep the format, just copy from the original page manually.

**Comment by matthiasdebernardini** [comment]
Hi not sure if its okay to post here, I can post it somewhere else. I am curious how to get this functionality in Safari? As in highlighting text with cmd+a then right clicking and allowing me to save it to a new quiver note. Is this possible yet?

**Comment by imcoddy** [comment]
Is it possible to add a shortcut for triggering the webclipper and save it to Quiver without mouse clicking?
Also it would be a lot convenient to include the source URL after saving it as a note :)

**Comment by purgat0ri** [comment]
Is a version for Safari on the way?

**Comment by ylian** [comment]
@purgat0ri Yes, Safari extension comes together with Quiver for macOS 3.2.5. Aiming to be released this week. The extension will also support clipping a selection, Markdown mode, clip as Markdown, etc. It will be available for Firefox too. All in this week.

**Comment by ylian** [comment]
Quiver Web Clipper v1.1.5 is now available on Firefox: https://addons.mozilla.org/en-US/firefox/addon/quiver-web-clipper/

The Chrome version (https://chrome.google.com/webstore/detail/quiver-web-clipper/hcnffmpopoelpggikahccdfenoobjigj) has also been updated to v1.1.5. Changes include:
* Markdown mode
* Clip as Markdown
* Clip a selection (use the context menu).

The Safari version is coming out this week as part of Quiver for macOS 3.2.5.

**Comment by purgat0ri** [comment]
Excellent! Thanks so much for this. At first I thought it didn't work, since clicking on 'Clip to Quiver' did not appear to do anything. After restarting Firefox, however, it all works perfectly!

**Comment by zshbleaker** [comment]
3 weeks passed and still no version 3.2.5 w/ Safari clipper extension, really need this.

**Comment by kimaldis** [comment]
Enable reader mode, select all or section, copy, paste

**Comment by ylian** [comment]
@zshbleaker Was fixing a critical bug. Finally fixed thanks to a user's help. 3.2.5 coming out in a couple days.

**Comment by ylian** [comment]
@zshbleaker 3.2.5 is released. After updating to 3.2.5, open Safari > Preferences > Extensions, and you will find Quiver Web Clipper as a Safari app extension.

**Comment by iiKulakov** [comment]
Not working on this page:

[https://medium.com/@derrickmehaffy/using-strapi-in-production-with-https-717af82d1445](url)

**Comment by zshbleaker** [comment]
@ylian Nice work bro, it works pretty fine. 👍

**Comment by girvo** [comment]
> @zshbleaker 3.2.5 is released. After updating to 3.2.5, open Safari > Preferences > Extensions, and you will find Quiver Web Clipper as a Safari app extension.

Any rough ideas why clicking the checkbox to enable the extension in Safari doesn't enable the extension?

If I uninstall the extension, how do I reinstall it?
